### ISSUE TITLE
Data channel fails to send packets at a frequently speed when works as a rtc server and connected by a browser.

### ISSUE THREAD

**Issue by HsuJv** [webrtc-rs/webrtc#360]
Hi all,

I'm trying to test the bandwidth of the Data Channel based on [your example](https://github.com/webrtc-rs/webrtc/blob/master/examples/examples/data-channels/data-channels.rs)

I didn't change too many codes except the data channel callbacks.
I tried to send as much as possible to the client when a data channel was created (will show in the codes later).
And with a controllable delay, I tried send frequency at10Hz, 20Hz, 25Hz, 40Hz, 50Hz...etc

The problem is, when it comes to over than 40Hz, the data received in the browser mismatches the data I'm trying to send.
With the send frequency set to 40Hz (65535 bytes per send call), I've got the throughput jitter sharply and the browser receives fewer packets than I sent.
![image](https://user-images.githubusercontent.com/14556195/206388653-daf098f7-14a4-470a-b109-7c140c38f0ee.png)
The gap increases when the send frequency becomes larger.
(Note here in the picture 192.168.1.2 is my webrtc server which runs the codes and sends data)

With the send frequency set to 25Hz (65535 bytes per send call as well),  I've got the throughput smoothly and the browser receives all the data I sent.
![image](https://user-images.githubusercontent.com/14556195/206389626-d6ba7250-ddbd-4d73-aedd-1960c2e817f3.png)

I've got no idea what is going wrong, please if anyone here can help?

P.S.
Below are my code snippets if they can be helpful.

I rewrite these codes
https://github.com/webrtc-rs/webrtc/blob/20b59b7a40663c7546b105d7cd698ca4655a1fc4/examples/examples/data-channels/data-channels.rs#L117-L155
to the following
```Rust
    // Register data channel creation handling
    peer_connection.on_data_channel(Box::new(move |d: Arc<RTCDataChannel>| {
        let d_label = d.label().to_owned();
        let d_id = d.id();
        println!("New DataChannel {} {}", d_label, d_id);

        // Register channel opening handling
        Box::pin(async move {
            use tokio::time;
            let d2 = Arc::clone(&d);
            let d_label2 = d_label.clone();
            let d_id2 = d_id;

            d.on_open(Box::new(move || {
                println!("Data channel '{}'-'{}' open.", d_label2, d_id2);
                let mut buf = Vec::with_capacity(65535);
                let mut current = time::Instant::now();
                let mut pkt_num = 0_usize;
                unsafe {
                    buf.set_len(65535);
                }

                Box::pin(async move {
                    while d2.send(&buf.to_vec().into()).await.is_ok() {
                        pkt_num  += 1;
                        time::sleep(time::Duration::from_millis(40)).await;
                        if current.elapsed().as_secs() > 1 {
                            println!("current send {} packets", pkt_num);
                            current = time::Instant::now();
                        }
                    }
                })
            }));

            // Register text message handling
            d.on_message(Box::new(move |_: DataChannelMessage| {
                Box::pin(async move {})
            }));

            d.on_close(Box::new(|| std::process::exit(0)));
        })
    }));
```

And my Javascript codes are quite simple as shown below
```Javascript
let pc = new RTCPeerConnection({
    iceServers: [
        {
            urls: 'stun:stun.l.google.com:19302'
        }
    ]
})

var recv_bytes = 0;
var pkts_num = 0;
let interval = null;

let sendChannel = pc.createDataChannel('foo')
sendChannel.onclose = () => {
    console.log('sendChannel has closed')
    clearInterval(interval);
}
sendChannel.onopen = () => {
    console.log('sendChannel has opened')
    interval = setInterval(() => {
        console.log("Recv " + pkts_num + " packets. Current rate: " + recv_bytes + " B/s");
        recv_bytes = 0;
    }, 1000);
}
sendChannel.onmessage = e => {
    recv_bytes += e.data.byteLength
    pkts_num += 1
    // sendChannel.send(e.data)
}

pc.oniceconnectionstatechange = e => console.log(pc.iceConnectionState)
pc.onnegotiationneeded = e =>
    pc.createOffer().then(d => {
        pc.setLocalDescription(d)
        fetch('/sdp', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            },
            body: "client=" + btoa(JSON.stringify(d))
        }).then(response => response.text().then(response => {
            pc.setRemoteDescription(new RTCSessionDescription(JSON.parse(atob(response))))
        }))
    }).catch(console.log)

```

**Comment by KillingSpark** [comment]
There is issue #101 which might be what you are running up against?

**Comment by HsuJv** [comment]
@KillingSpark ,
Thanks for the information!

I tried the example in the issue, and it works well with 1024 bytes per send call.
But when I increase the send data to 16384 bytes per call, it still fails.
The `buffered_amount` of the data channel just doesn't get down unless I increase the sleep time to 10ms (I tried 10 micros, 100 micros, and 1000 micros which didn't help).

My question here is what exactly blocks the `buffered_amount` from getting down

![image](https://user-images.githubusercontent.com/14556195/206400027-cb5ecee2-b04c-4c40-9ed7-8da5a7e793c1.png)
![image](https://user-images.githubusercontent.com/14556195/206401396-24db0d6c-4331-418e-b018-ba3b72f0e099.png)

**Comment by KillingSpark** [comment]
I can only guess, I don't have deep knowledge on this library (I am just a user with a few small contributions :)). But the default for datachannels is probably reliable transport, so if the bandwidth is used to the maximum, it will start buffering data, just like tcp would.

Just out of interest, does this change if you open a datachannel with an unreliable transport? Should be a setting on the javascript API somewhere.

**Comment by HsuJv** [comment]
I've got no idea but I'll try it out later.
Anyway thanks for the help.
Let's leave the issue here to see if anyone else would like to answer.

**Comment by HsuJv** [comment]
It seems something is wrong within the `SCTP` 

I just wrote a simple POC to do a throughput test
```Rust
use clap::{App, AppSettings, Arg};
use std::io::Write;
use std::sync::Arc;
use std::time::Duration;
use tokio::net::UdpSocket;
use util::{conn::conn_disconnected_packet::DisconnectedPacketConn, Conn};
use webrtc_sctp::association::*;
use webrtc_sctp::chunk::chunk_payload_data::PayloadProtocolIdentifier;
use webrtc_sctp::stream::*;
use webrtc_sctp::Error;

#[tokio::main]
async fn main() -> Result<(), Error> {
    env_logger::Builder::new()
        .format(|buf, record| {
            writeln!(
                buf,
                "{}:{} [{}] {} - {}",
                record.file().unwrap_or("unknown"),
                record.line().unwrap_or(0),
                record.level(),
                chrono::Local::now().format("%H:%M:%S.%6f"),
                record.args()
            )
        })
        .filter(None, log::LevelFilter::Warn)
        .init();

    let mut app = App::new("SCTP Throughput")
        .version("0.1.0")
        .about("An example of SCTP Server")
        .setting(AppSettings::DeriveDisplayOrder)
        .setting(AppSettings::SubcommandsNegateReqs)
        .arg(
            Arg::with_name("FULLHELP")
                .help("Prints more detailed help information")
                .long("fullhelp"),
        )
        .arg(
            Arg::with_name("port")
                .required_unless("FULLHELP")
                .takes_value(true)
                .long("port")
                .help("use port ."),
        );

    let matches = app.clone().get_matches();

    if matches.is_present("FULLHELP") {
        app.print_long_help().unwrap();
        std::process::exit(0);
    }

    let port1 = matches.value_of("port").unwrap().to_owned();
    let port2 = port1.clone();

    std::thread::spawn(|| {
        tokio::runtime::Runtime::new()
            .unwrap()
            .block_on(async move {
                let conn = DisconnectedPacketConn::new(Arc::new(
                    UdpSocket::bind(format!("127.0.0.1:{}", port1))
                        .await
                        .unwrap(),
                ));
                println!("listening {}...", conn.local_addr().unwrap());

                let config = Config {
                    net_conn: Arc::new(conn),
                    max_receive_buffer_size: 0,
                    max_message_size: 0,
                    name: "server".to_owned(),
                };
                let a = Association::server(config).await?;
                println!("created a server");

                let stream = a.accept_stream().await.unwrap();
                println!("accepted a stream");

                // set unordered = true and 10ms treshold for dropping packets
                stream.set_reliability_params(true, ReliabilityType::Timed, 10);

                let mut buff = vec![0u8; 65535];
                let mut recv = 0;
                let mut pkt_num = 0;
                let mut now = tokio::time::Instant::now();
                while let Ok(n) = stream.read(&mut buff).await {
                    recv += n;
                    if n != 0 {
                        pkt_num += 1;
                    }
                    if now.elapsed().as_secs() == 1 {
                        println!("Throughput: {} Bytes/s, {} pkts", recv, pkt_num);
                        now = tokio::time::Instant::now();
                        recv = 0;
                        pkt_num = 0;
                    }
                }
                Result::<(), Error>::Ok(())
            })
    });

    std::thread::spawn(|| {
        tokio::runtime::Runtime::new()
            .unwrap()
            .block_on(async move {
                let conn = Arc::new(UdpSocket::bind("0.0.0.0:0").await.unwrap());
                conn.connect(format!("127.0.0.1:{}", port2)).await.unwrap();
                println!("connecting {}..", format!("127.0.0.1:{}", port2));

                let config = Config {
                    net_conn: conn,
                    max_receive_buffer_size: 0,
                    max_message_size: 0,
                    name: "client".to_owned(),
                };
                let a = Association::client(config).await.unwrap();
                println!("created a client");

                let stream = a
                    .open_stream(0, PayloadProtocolIdentifier::Binary)
                    .await
                    .unwrap();
                println!("opened a stream");

                let mut buf = Vec::with_capacity(65535);
                unsafe {
                    buf.set_len(65535);
                }

                let mut now = tokio::time::Instant::now();
                let mut pkt_num = 0;
                while stream.write(&buf.clone().into()).is_ok() {
                    pkt_num += 1;
                    if now.elapsed().as_secs() == 1 {
                        println!("Send {} pkts", pkt_num);
                        now = tokio::time::Instant::now();
                        pkt_num = 0;
                    }
                }
                Result::<(), Error>::Ok(())
            })
    });
    loop {}
}
```

And got logs like 
```plain
listening 127.0.0.1:8888...
connecting 127.0.0.1:8888..
created a server
created a client
opened a stream
accepted a stream
Send 20245 pkts
Throughput: 4980660 Bytes/s, 76 pkts
Send 20135 pkts
Throughput: 5046195 Bytes/s, 77 pkts
Send 19996 pkts
Throughput: 5111730 Bytes/s, 78 pkts
```
whatever the `set_reliability_params` is `true` or `false`

**Comment by HsuJv** [comment]
My UDP snippets were like this
```Rust
fn main() -> Result<(), Error> {
    env_logger::Builder::new()
        .format(|buf, record| {
            writeln!(
                buf,
                "{}:{} [{}] {} - {}",
                record.file().unwrap_or("unknown"),
                record.line().unwrap_or(0),
                record.level(),
                chrono::Local::now().format("%H:%M:%S.%6f"),
                record.args()
            )
        })
        .filter(None, log::LevelFilter::Warn)
        .init();

    let mut app = App::new("SCTP Throughput")
        .version("0.1.0")
        .about("An example of SCTP Server")
        .setting(AppSettings::DeriveDisplayOrder)
        .setting(AppSettings::SubcommandsNegateReqs)
        .arg(
            Arg::with_name("FULLHELP")
                .help("Prints more detailed help information")
                .long("fullhelp"),
        )
        .arg(
            Arg::with_name("port")
                .required_unless("FULLHELP")
                .takes_value(true)
                .long("port")
                .help("use port ."),
        );

    let matches = app.clone().get_matches();

    if matches.is_present("FULLHELP") {
        app.print_long_help().unwrap();
        std::process::exit(0);
    }

    let port1 = matches.value_of("port").unwrap().to_owned();
    let port2 = port1.clone();

    std::thread::spawn(|| {
        tokio::runtime::Runtime::new()
            .unwrap()
            .block_on(async move {
                let conn = UdpSocket::bind(format!("127.0.0.1:{}", port1))
                    .await
                    .unwrap();
                println!("listening {}", format!("127.0.0.1:{}", port1));

                let mut buff = vec![0u8; 65535];
                let mut recv = 0;
                let mut pkt_num = 0;
                let mut now = tokio::time::Instant::now();
                while let Ok(n) = conn.recv(&mut buff).await {
                    recv += n;
                    if n != 0 {
                        pkt_num += 1;
                    }
                    if now.elapsed().as_secs() == 1 {
                        println!("Throughput: {} Bytes/s, {} pkts", recv, pkt_num);
                        now = tokio::time::Instant::now();
                        recv = 0;
                        pkt_num = 0;
                    }
                }
                Result::<(), Error>::Ok(())
            })
    });

    std::thread::spawn(|| {
        tokio::runtime::Runtime::new()
            .unwrap()
            .block_on(async move {
                let conn = UdpSocket::bind("0.0.0.0:0").await.unwrap();
                println!("Connect to {}", format!("127.0.0.1:{}", port2));
                conn.connect(format!("127.0.0.1:{}", port2)).await.unwrap();
                println!("Connected");

                let mut buf = Vec::with_capacity(16384);
                unsafe {
                    buf.set_len(16384);
                }

                let mut now = tokio::time::Instant::now();
                let mut pkt_num = 0;
                while conn.send(&buf).await.is_ok() {
                    pkt_num += 1;
                    if now.elapsed().as_secs() == 1 {
                        println!("Send {} pkts", pkt_num);
                        now = tokio::time::Instant::now();
                        pkt_num = 0;
                    }
                }
                Result::<(), Error>::Ok(())
            })
    });
    loop {}
}
```
With this I think we can rule out the possibility that tokio goes wrong.
Hope this would be helpful for you.
I will do some more research when I have time.

BRs.

**Comment by KillingSpark** [comment]
Ok so I did some more digging and need to take back what I wrote earlier. It is NOT the received queue of the socket filling up.

Two things i found:
1. The sctp layer decides to limit how many packets are sent
2. The pending queue grows indefinitely (partly because of 1. but this should apply back pressure to the code actually sending packets)

So there are two separate problems at play here:
1. SCTP is limiting me to about 40MBit/s in release mode. This seems way smaller than it should be, considering the speeds that raw UDP can provide.
2. The pending queue does not apply backpressure, causing these confusing logs. If this backpressure had been applied it would have been obvious that the sender is slow and not the reader. It's also important in general that this library doesn't just fill up all the memory just because the application is writing faster than the connection is sending data.

**Comment by KillingSpark** [comment]
More info: it seems like the ACK mechanism is borked:

I just did dumb println! debugging in `association_internal.rs`
* "Handle" => "handle_data"
*  "Send" => "pop_sending_data_chunks_to_send"
* "ACK" => "create_selective_ack_chunk"
* "ACKRCV => "process_selective_ack" 

```
Handle TSN: 438973196
Send   TSN: 438973199
Handle TSN: 438973197
Send   TSN: 438973200
Handle TSN: 438973198
ACK    TSN: 438973198
Send   TSN: 438973201
ACKRCV TSN: 438972331
Send   TSN: 438973202
Handle TSN: 438973199
Handle TSN: 438973200
Send   TSN: 438973203
Handle TSN: 438973201
ACK    TSN: 438973201
Send   TSN: 438973204
ACKRCV TSN: 438972334
Handle TSN: 438973202
Handle TSN: 438973203
Handle TSN: 438973204
```

Note that Send/Handle/ACK seem to match pretty well while ACKRCV is laggin behind significantly. So either

1. The ACK is sent with a big delay
2. The ACK is processed with a big delay

I'll see if I can determin which it is...

**Comment by KillingSpark** [comment]
Yup it's lock contention. Specifically on the Mutex around the InternalAssociation struct.

Occupied time in readloop measures time used to process an SACK and in writeloop it's gathering packets to send.

Taking time is just measuring the time it took to actually lock the mutex before doing the above operation.

It seems pretty clear to me that the SACK processing is stalled by the lock contention. The solution to this is probably non-trivial. Maybe instead of making the writeloop gather packets to send, the packets should be put into a channel either when they are queued and can be sent immediatly within the rwnd or when a SACK arrives that increases the rwnd?

```
Writeloop lock occupied: 1431us
readloop lock occupied:  43us
Writeloop lock occupied: 38us
Writeloop lock occupied: 1us
Writeloop lock occupied: 1548us
readloop lock occupied:  46us
Writeloop lock occupied: 17us
Writeloop lock occupied: 0us
Writeloop lock occupied: 1498us
readloop lock occupied:  47us
Writeloop lock occupied: 66us
Writeloop lock occupied: 1us
Writeloop lock occupied: 1412us
readloop lock occupied:  43us
Writeloop lock occupied: 19us
Writeloop lock occupied: 0us
Writeloop lock occupied: 1426us
readloop lock occupied:  43us
Writeloop lock occupied: 19us
Writeloop lock occupied: 1us
Writeloop lock occupied: 1552us
readloop lock occupied:  68us
Writeloop lock occupied: 19us
Writeloop lock occupied: 0us
Writeloop lock occupied: 1641us
```

```
Taking Writeloop lock: 0us
Taking Writeloop lock: 0us
Taking readloop lock:  791us
Taking Writeloop lock: 66us
Taking Writeloop lock: 0us
Taking Writeloop lock: 0us
Taking readloop lock:  825us
Taking Writeloop lock: 65us
Taking Writeloop lock: 0us
Taking Writeloop lock: 0us
Taking readloop lock:  870us
Taking Writeloop lock: 68us
Taking Writeloop lock: 0us
Taking Writeloop lock: 0us
Taking readloop lock:  861us
Taking Writeloop lock: 71us
Taking Writeloop lock: 0us
Taking Writeloop lock: 0us
Taking readloop lock:  846us
Taking Writeloop lock: 71us
Taking Writeloop lock: 0us
Taking Writeloop lock: 0us
Taking readloop lock:  863us
Taking Writeloop lock: 68us
Taking Writeloop lock: 0us
Taking Writeloop lock: 0us
Taking readloop lock:  908us
Taking Writeloop lock: 69us
Taking Writeloop lock: 0us
Taking Writeloop lock: 0us
Taking readloop lock:  810us
Taking Writeloop lock: 64us
Taking Writeloop lock: 0us
Taking Writeloop lock: 0us
Taking readloop lock:  885us
Taking Writeloop lock: 72us
Taking Writeloop lock: 0us
Taking Writeloop lock: 0us

```

**Comment by KillingSpark** [comment]
Ok so my suspicion was right. Pulling the marshalling our from under the lock reduces the time the mutex is locked in total drastically... but: this doesn't really help because the tokio scheduler thinks it is a good idea to have both the write and read loop on the same thread.

```
sender Try taking readloop lock on ThreadId(11)           <---- Start waiting on lock in read_loop
sender Taking Writeloop lock: 34us
sender Writeloop lock occupied: 87us                      <---- Drop lock in write_loop
sender Run write body on ThreadId(11)
sender Writeloop body took: 838us                         <---- write_loop still occupies thread
sender Try taking writeloop lock
sender Taking readloop lock:  974us                       <---- read_loop gets access to lock
sender readloop lock occupied:  5us
sender Try taking readloop lock on ThreadId(11)
... repeat
```

As far as I can see the two loops need to run on different threads to maximize throughput.

**Comment by xnorpx** [comment]
did you see improvement in throughput?

**Comment by HsuJv** [comment]
Awesome work!
I've been a bit busy for a while and haven't had time to research this issue.
From 14MB/s to 95MB, it looks like quite a significant improvement
But we still seem to have a long way to go compared to webrtc-go.
Anyway, I appreciate your enthusiastic help on this issue, the problem I am currently facing seems to have been solved in this way for the most part.
Hope this change can be merged into releases

BRs

**Comment by KillingSpark** [comment]
Even before the changes I could get ~70MB/s in release mode, but the changes I did to get to 95 where pretty small, I'd guess there is even more potential for optimization.

**Comment by k0nserv** [comment]
Excellent research, sounds like your changes are promising @KillingSpark. Please roll it up into PRs for review

**Comment by KillingSpark** [comment]
The three PRs comgined have this effect for me:

### Current master
```
Throughput: 58522755 Bytes/s, 893 pkts, 893 loops
Send 89902 pkts
Throughput: 60030060 Bytes/s, 916 pkts, 916 loops
Send 90575 pkts
Throughput: 60161130 Bytes/s, 918 pkts, 918 loops
Send 89910 pkts
Throughput: 59374710 Bytes/s, 906 pkts, 906 loops
Send 89765 pkts
Throughput: 58850430 Bytes/s, 898 pkts, 898 loops
```

### All three PRs combined
```
Throughput: 103938510 Bytes/s, 1586 pkts, 1586 loops
Send 83505 pkts
Throughput: 139196340 Bytes/s, 2124 pkts, 2124 loops
Send 80997 pkts
Throughput: 136247265 Bytes/s, 2079 pkts, 2079 loops
Send 83435 pkts
Throughput: 136706010 Bytes/s, 2086 pkts, 2086 loops
Send 84264 pkts
Throughput: 140244900 Bytes/s, 2140 pkts, 2140 loops
```

Still not at pion levels but the next big botlleneck that exists now seems to be the CRC32 impementation.

**Comment by k0nserv** [comment]
@KillingSpark is there a specific order of merging these that would be simpler than the others?

**Comment by k0nserv** [comment]
Also, can you please add changelog entries in yourPRs?

**Comment by KillingSpark** [comment]
Will do!

Order of merging shouldn't matter they should be independent.

But hold off from merging, I have one more idea on the write_loop behaviour I want to try.

**Comment by KillingSpark** [comment]
Ok, changelogs has been done, my idea didn't work out, so you can merge if you want to

**Comment by HsuJv** [comment]
I tested with the release build in my WSL2 Arch (another pc), and the throughput of the SCTP increased from 140-200 MBps to 400-500 MBps, awesome!
But it seems the send queue still increases infinitely. Should we add some back pressure when it comes to fast-slow issues?

**Comment by KillingSpark** [comment]
Definitely should do that. I have some ideas for that as well (and prototyping shows it increases throughput even more) but I wanted to get these PRs merged before starting another one :)

**Comment by k0nserv** [comment]
@melekes would you mind helping with some reviews, you have a more subject matter expertise than me in this area

**Comment by KillingSpark** [comment]
I am a bit lost here to be honest.

If I use the optimized sctp marshalling the performance just tanks on my linux PC and I cannot figure out why. On the Mac this improves performance a lot.

But it isn't really about the scheduling either as far as I can tell.

tokio-console also doesn't reveal any glaring issues.

Does anyone have ideas? HsuJv mentioned that in WSL this brought big improvements too, so windows and mac seem to see an improvement but it seems very weird that on linux this would cause such a drastic performance  decrease... In the end it is a pretty straight forward optimization...

**Comment by HsuJv** [comment]
On my ubuntu 20.04 server (i7-7700 16G RAM)
The throughput increased from 200-400MB to around 700MB, but it jitters very sharply like this
```plain
istening 127.0.0.1:8888...
connecting 127.0.0.1:8888..
created a server
created a client
opened a stream
accepted a stream
Throughput: 15204120 Bytes/s, 232 pkts, 232 loops
Send 15505 pkts
Throughput: 127989855 Bytes/s, 1953 pkts, 1953 loops
Send 2089 pkts
Throughput: 23395995 Bytes/s, 357 pkts, 357 loops
Send 223 pkts
Throughput: 135788520 Bytes/s, 2072 pkts, 2072 loops
Send 2069 pkts
Throughput: 17170170 Bytes/s, 262 pkts, 262 loops
Send 265 pkts
Throughput: 43777380 Bytes/s, 668 pkts, 668 loops
Send 668 pkts
Throughput: 51117300 Bytes/s, 780 pkts, 780 loops
Send 780 pkts
Throughput: 132315165 Bytes/s, 2019 pkts, 2019 loops
Send 2015 pkts
Throughput: 139786155 Bytes/s, 2133 pkts, 2133 loops
Send 2137 pkts
Throughput: 3604425 Bytes/s, 55 pkts, 55 loops
Send 55 pkts
Throughput: 30801450 Bytes/s, 470 pkts, 470 loops
Send 470 pkts
Throughput: 107608470 Bytes/s, 1642 pkts, 1642 loops
Send 2063 pkts
Throughput: 88079040 Bytes/s, 1344 pkts, 1344 loops
Send 923 pkts

```

**Comment by KillingSpark** [comment]
This is somewhat consistent with what I see on my machine though my spikes are much rarer. I am on a Ryzen 3800. 

Does Tokio employe different scheduling depending on the OS?

**Comment by KillingSpark** [comment]
Fixed it. @HsuJv could you try the current state of the merged branch one more time just to confirm this?

**Comment by HsuJv** [comment]
Yes it gets smoothly at around 130MB
BTW, In my last comment, the number was 20-40 MB & 70 MB. I did a wrong calculation.
```plain
accepted a stream
Throughput: 262140 Bytes/s, 4 pkts, 4 loops
Send 15279 pkts
Throughput: 95550030 Bytes/s, 1458 pkts, 1458 loops
Send 1850 pkts
Throughput: 130152510 Bytes/s, 1986 pkts, 1986 loops
Send 1981 pkts
Throughput: 131594280 Bytes/s, 2008 pkts, 2008 loops
Send 2008 pkts
Throughput: 132315165 Bytes/s, 2019 pkts, 2019 loops
Send 2030 pkts
Throughput: 131528745 Bytes/s, 2007 pkts, 2007 loops
Send 1995 pkts
Throughput: 131070000 Bytes/s, 2000 pkts, 2000 loops
Send 1996 pkts
Throughput: 129824835 Bytes/s, 1981 pkts, 1981 loops
```
(My server always runs some heavy tasks, so it is expected that the performance is lower than my WSL)

**Comment by melekes** [comment]
@HsuJv @KillingSpark thanks for looking into this 👍 fantastic work

**Comment by k0nserv** [comment]
Btw, worth keeping in mind how this work intersects with https://github.com/webrtc-rs/webrtc/issues/136 and a whether this potential move to a sans-IO version of SCTP helps improve the performance problems.

In any case, excellent work on this 👍🏼

**Comment by k0nserv** [comment]
We were discussing it Discord the other day and would like to move forward with switching out the implementation of SCTP with it. However, no one is committing to do that work at the moment so it would only happen when someone finds the time for it
